### PR TITLE
Initial popgen page

### PR DESCRIPTION
**Pull request by benjeffery**
Start at #43 
What quantities do we want here? For anything interesting, we likely need to allow the user to specify sample sets, maybe by metadata key?

https://github.com/tskit-dev/tsinfer-qc/assets/8552/e8f92922-7693-4b5f-8ac7-c80ccf561dc7

### CODE REVIEW

**Comment by jeromekelleher** [comment]
This is a great start! @savitakartik will go through a bit later and comment

File: pages/popgen.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    tajimas_curve = hv.Curve((x_axis, tajimas), "Genomic position", "Tajimas D").opts(
+        interpolation="steps-post"
+    )
+
+    def set_callbacks(fig, index, data):
+        source = ColumnDataSource(dict(index=index, x=data))
+
+        def hook(hv_plot, element):
+            fig = hv_plot.handles["plot"]
+            callback = CustomJS(
```
Comment (jeromekelleher) [inline pages/popgen.py:33]: What's this JS code for?

**Comment by jeromekelleher** [comment]
Looks really nice - would be nice to not need embedded JS code though, as it's quite likely to get broken. Unless it's really doing something crucial I would vote for not having it?

**Comment by jeromekelleher** [comment]
Re the sample sets stuff - I think we can punt this down the road. Probably we'll offer the option of comparing different populations or metadata keys (as you say) but don't need it initially.

File: pages/popgen.py (hunk 1 above)
Comment (benjeffery) [reply pages/popgen.py:33]: This JS restricts panning and zooming to be horizontal. After any movement it adjusts the y axis range to be the extent of the visible data. You can see it working in the video at the top of the PR. Without this the mousewheel zooms in 2D for a frustrating experience!

The API it is using (`cb_obj`) is documented here: https://docs.bokeh.org/en/2.4.1/docs/user_guide/interaction/callbacks.html#customjs-for-user-interaction-events

**Comment by benjeffery** [comment]
> * mode=node doesn't make sense (and leads to crashes), we shouldn't provide it as an option
> * mode="branch" seems like a better default as it tells us useful things about the trees

Sure, makes sense.

> * The "windows" slider doesn't work very well. Maybe a "num_windows" option would be more useful across different species?

Yep, can do.

> * It's not obvious how "tree based windows" interacts with the slider

I need to remove the slider if tree-based is selected as it is ignored.

File: pages/popgen.py (hunk 1 above)
Comment (jeromekelleher) [reply pages/popgen.py:33]: I see, sounds good. We'll probably want to do this across a few different plots then, so how about we pull it out for now and create an issue to track?

**Comment by jeromekelleher** [comment]
What do you think @savitakartik?

**Comment by savitakartik** [comment]
Just tried it out. Looks great. I just had one small comment - I get an numpy error when I try running the app with a tiny ts.

File: pages/popgen.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+def make_one_way_stats(ts, windows_trees, windows_count, span_normalise, mode):
+    mode = mode.lower().split()[0]
+    windows_count = windows_count
+    if windows_trees:
+        windows = "trees"
+        x_axis = ts.breakpoints(as_array=True)[:-1]
+    else:
+        windows = np.concatenate(
+            (
+                np.arange(0, ts.sequence_length, ts.sequence_length // windows_count),
```
Comment (savitakartik) [inline pages/popgen.py:19]: I get this error when I try to use a tiny ts with 2 trees:
ZeroDivisionError: float division by zero

**Comment by savitakartik** [comment]
I love this page - would make things like checking for well-known selection signals so easy.

**Comment by jeromekelleher** [comment]
Can you rebase and add this page to the ``pages`` tests please @benjeffery ? That should provoke the error @savitakartik found above.

File: pages/popgen.py (hunk 2 above)
Comment (benjeffery) [reply pages/popgen.py:19]: Ah, good catch.

**Comment by benjeffery** [comment]
Rebased, added to tests and JS removed. https://github.com/tskit-dev/tsinfer-qc/issues/79 filed.

File: pages/popgen.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+import holoviews as hv
+import numpy as np
+import panel as pn
+from holoviews import opts
+
+
+def make_one_way_stats(ts, windows_trees, windows_count, span_normalise, mode):
+    mode = mode.lower().split()[0]
+    windows_count = windows_count
+    print(windows_count, ts.sequence_length)
```
Comment (jeromekelleher) [inline pages/popgen.py:10]: stray print

**Review by jeromekelleher — APPROVED** [review]
LGTM - one stray print

File: pages/popgen.py (hunk 3 above)
Comment (benjeffery) [reply pages/popgen.py]: Dang it! Fixed.

**Pull request merged**